File: memo.java
// ======= Integer =======
int max = Integer.MAX_VALUE; // java integer's max value
long a = (long)3; // convert int to long
rand.nextInt(bound); // return random element in bound
Integer.parseInt(str); // convert string to int
Math.round(d * 1000) / 1000; // 3 decimal point int

// ======= Double =======
Double b = 3; // init Double from int

// ======= String =======
String str = "abc";
str.length(); // size of string
char[] arr = s.toCharArray(); //string to char array
s.substring(i) + s.substring(0, i); // substring
s1.compareTo(s2); // compare 2 strings, 0 means equal, <0 means s1 < s2, >0 means s1 > s2
s1.equals(str); // check if s1 equal to str
s.charAt(0); // return char at index
int charA = 'a'; // convert char to int
(int)'a'; // convert char to int
(char)next; // convert int to char
str.indexOf('a'); // first index appear of char a
s.trim().split("\\s+"); //split by multiple whitespace
String.join(" ", str); // join string from array
String.valueOf(3); // "3" convert int to string
Character.isUpperCase(word.charAt(i)); // check if a character is upper/capital letter
// init repeating character string
char[] chars = new char[len];
Arrays.fill(chars, ch);
String s = new String(chars);


// ======= StringBuilder =======
StringBuilder sb = new StringBuilder();
sb.append(s.substring(i+1, end+1)).append(" "); // append string
sb.toString(); // convert stringbuilder to string, can also convert substring to string
sb.reverse(); // reverse string by stringBuilder
sb.charAt(i); // access stringBuilder character
sb.insert(0, char); // insert char at 0 index
sb.deleteCharAt(0); // delete character at index 0

// ======= StringBuilder =======
Character.isDigit(c); // check if c is digit number
(s.charAt(i)-'0'); // convert char to int

// ======= Arrays =======　
int[] dir = new int[] {0, 1, 0, -1, 0}; // array initialization
Arrays.sort(arr); // sort array
Arrays.fill(arr, 1); // fill array
Arrays.copyOf(arr, newSize); // return a copy of arr
int[] arr = new int[3];
arr.length; // return arr's size
arr.isEmpty(); // check arr is empty or not
arr.add(ele); // array list add element
arr.stream().mapToInt(Integer::intValue).toArray(); // convert arrayList to array
Arrays.asList(nums); // convert array to list
Collections.swap(arr, i, j); // swap element in collections
int index = Arrays.binarySearch(nums, target); // binary search target, if exist, return correct index, if not return "-(should be inserted index)"
Arrays.copyOfRange(nums, n-k, n); // get range of element in array
System.arraycopy(lastK, 0, nums, 0, k); // copy lastK's k elements to nums[0..<k]
Collections.reverse(Arrays.asList(str)) // reverse array
Arrays.sort(points, new Comparator<int[]>() { // sort array
    public int compare(int[] a, int[] b) {
        return Integer.compare(a[1],b[1]);
    }
}) ;
Arrays.sort(points, (p1, p2) -> {
    if (p1[0] == p2[0]) {
        return Integer.compare(p1[1], p2[1]);
    } else {
        return Integer.compare(p1[0], p2[0]);
    }
});
boolean[] b = new boolean[3]; // init boolean array

// ======= ArrayList =======　
List<Integer> arr = new ArrayList<Integer>();
arr.get(0); // array list get element in the index
arr.add(0); // array list add element
arr.size(); // array list size
arr.subList(3, 4); // sublist of [3..<4]
Arrays.asList(1, 2, 3); // convert array to list
arr.iterator().next(); // iterate over array list
numsList.add(0, 1); // add 1 at 0 index
arr.remove(arr.size()-1); // remove last element int the list
List<Plant> copy = new ArrayList<>(list); // copy list

// ======= Queue =======　
Queue<TreeNode> q = new LinkedList<>(); // offer() insert element, poll return and remove first, peek return but not remove
q.size(); // get queue size
q.offer(ele); // add element to queue
q.poll(); //返回第一个元素，并在队列中删除
q.peek(); //返回第一个元素
Queue<Integer> queue = new PriorityQueue(Collections.reverseOrder()); // priorityQueue, heap

// ======= Stack =======　
Stack<Integer> stack = new Stack<>();
stack.peek(); // return stack first object
stack.pop(); // remove first in stack and return it
stack.push(ele); // push element into stack
stack.empty(); // check empty
Deque<Pair<TreeNode, Integer>> stack = new ArrayDeque(); // use ArrayDeque to init stack
stack.isEmpty(); // check empty

// ======= Set =======
Set<Integer> set = new HashSet<Integer>();
set.size(); // return size of set
set.contains(2); // check if hashset contains element

// ======= Map =======
Map<Integer, Integer>[] f = new Map[n]; // array of dictionary/map
map.getOrDefault(c, 0); // if key c exist, return value of c, otherwise, return 0
map.keySet(); // get map's keys
map.put(c, 0); // set key c's value as 0
map.remove(key); // remove key
map.computeIfAbsent(arr[i], v -> new LinkedList<>()).add(i); // give default value, then update the value
map.get(arr[cur]).clear(); // clear key
map.containsKey(2); // check if key exists in map
Map.Entry<Integer, Integer> entry: map.entrySet(); // return entry, can use getKey() and getValue() from it later

// ======= Math =======
Math.abs(-3); // get absolute value of one element
Math.sqrt(4); // sqrt value
Math.cell(0.3); // ceil number
(int)Math.floor(0.4); // floor number and convert to int, return 0

// ======= Random =======
Random rand = new Random();
rand.nextInt(n); // return integer in [0..<n-1]

// ======= Pair =======
Pair<String, Integer> person = new Pair<>("Sajal", 12);

File: memo.py
# int char
chr(97) # int to char 'a'
ord('a') # 97 char to int
int("12") # convert string to int
str(1) # convert int to string

sys.maxsize # int64 max
-sys.maxsize - 1 # int64 min
2**31-1 # int32 max
-2**31-1 #int32 min

# String
s.split(' ') # split string by space\
c = 'a'
c.isupper() # is upper case
c1.islower()
c.isnumeric()
c1.lower() # convert to lower char
str.join(sequence) # join sequence by str, could use to join char array
''.join(list(reversed(word))) # reverse string
str[::-1] # reverse string
str[i:j:] # return substring of s[i..<j]
max(str) # return its maximum digit, if str = "32", it will return 3
res = res[:-1] # drop last character

# List / array
# https://docs.python.org/3/tutorial/introduction.html#lists
len(array) # length of array
dp = [[0 for i in range(n)] for j in range(n)] # 2D array init
list.pop(i) # remove ith index element from list
list.insert(index, element) # insert element at index
pop(-1) # remove last element
arr.reverse() # reverse array in place
a, b = b, a # swap two element
nums[index+1:len(nums)] = nums[index+1:len(nums)][::-1] # reverse partial array
list.copy() # make copy of list without reference
bisect.bisect_left(a, x, lo=0, hi=len(a), *, key=None) # binary search, locate insertion point at left

# heap
import heapq
heapq.heapify(list) # convert list to heap (min heap), heappop, heappush,
heappush(list, 1) # push element in heap
heappop(list) # pop element from heap
list[0] # always return the smallest element
A = [-1 * a for a in A] # max heap
class Node(object):
    def __init__(self, val: int):
        self.val = val

    def __repr__(self):
        return f'Node value: {self.val}'

    def __lt__(self, other):
        return self.val < other.val

heap = [Node(2), Node(0), Node(1), Node(4), Node(2)]
heapq.heapify(heap)
print(heap) # custom heap

# Stack
stack = [(root, root.val)] # init stack with pair of <TreeNode, Int>
while stack # can use to check stack is empty or not
stack.pop() # use to pop stack value
stack.append((node, 23)) # push element into stack

# Set
startSet = set([0])
startSet.add(3) # add into set
startSet.discard(3) # if element exist in set, remove, otherwise do nothing
3 in startSet # check if element in set
if len(startSet) > len(endSet):
    startSet, endSet = endSet, startSet # quick switch element

# map
graph = {}
for i in range(n):
    if arr[i] in graph:
        graph[arr[i]].append(i)
    else:
        graph[arr[i]] = [i]
graph[arr[cur]].clear() # remove key from graph
graph.keys() # get all keys from graph
graph.values() # get all values from graph
graph.items() # get (key, val) pair
numMap[numSum] = numMap.get(numSum, 0) + 1 # get or default
if key in mydict # check key exist or not
Counter(nums) # convert nums to counter
sortedFreq = sorted(freq.items(), key=lambda x:(-x[1],x[0])) # sort map by descending value, ascending key
graph.pop(key) # remove key from graph

# others
range(1, 10) # range of 1..<10
range(0, -10, -1) # [0, -1, -2, -3, -4, -5, -6, -7, -8, -9] last one is step
a // 2 # 取整除 - 向下取接近商的整数 9//2 is 4
random.randint(0, 3) # random int of 0...3
@lru_cache(None) # cache function
"变量1" if a>b else "变量2"
math.ceil(0.1) # ceil math function
round(x, 2) # round number to only leave 2 decimal points
nonlocal res, a #  bind to the nearest non-global variable also called res, a
hA is hB # check if two object have same reference

# sort by comparator
def compare(p1, p2) -> bool:
    return p1[1] - p2[1]
points = sorted(points, key=cmp_to_key(compare))
points.sort(key = lambda x : x[1])
satisfaction.sort(key = cmp_to_key(lambda s1, s2 : s2-s1))

# map list element's names
list(map(lambda x: x[0], weaker))
